libnetwork/osl/interface_linux.go

```diff
@@ -276,22 +323,246 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+// disconnect or container stop. A new MAC address may have been generated, so
+// this is necessary to avoid packets sent to the old MAC address getting dropped
+// until the ARP/Neighbour cache entries expire.
+//
+// Note that the kernel's arp_notify sysctl setting is not respected.
+func (n *Namespace) advertiseAddrs(ctx context.Context, ifIndex int, i *Interface, nlh nlwrap.Handle) error {
+	mac := i.MacAddress()
+	ctx = log.WithLogger(ctx, log.G(ctx).WithFields(log.Fields{
+		"iface": i.dstName,
+		"ifi":   ifIndex,
+		"mac":   mac.String(),
+		"ip4":   i.Address(),
+		"ip6":   i.AddressIPv6(),
+	}))
+
+	if i.address == nil && i.addressIPv6 == nil {
+		// Nothing to do - for example, a bridge with no configured addresses.
+		log.G(ctx).Debug("No IP addresses to advertise")
+		return nil
+	}
+	if mac == nil {
+		// Nothing to do - for example, a layer-3 ipvlan.
+		log.G(ctx).Debug("No MAC address to advertise")
+		return nil
+	}
+	if i.advertiseAddrNMsgs == 0 {
+		log.G(ctx).Debug("Unsolicited ARP/NA is disabled")
+		return nil
+	}
+
+	arpSender, naSender := n.prepAdvertiseAddrs(ctx, i, ifIndex)
+	if arpSender == nil && naSender == nil {
+		return nil
+	}
+	cleanup := func() {
+		if arpSender != nil {
+			arpSender.Close()
+		}
+		if naSender != nil {
+			naSender.Close()
+		}
+	}
+	stillSending := false
+	defer func() {
+		if !stillSending {
+			cleanup()
+		}
+	}()
+
+	send := func(ctx context.Context, nSent int) error {
+		link, err := nlh.LinkByIndex(ifIndex)
+		if err != nil {
+			return fmt.Errorf("failed to refresh link attributes: %w", err)
+		}
+		if curMAC := link.Attrs().HardwareAddr; slices.Compare(curMAC, mac) != 0 {
+			log.G(ctx).WithFields(log.Fields{"newMAC": curMAC.String()}).Warn("MAC address changed")
+			return fmt.Errorf("MAC address changed, got %s, expected %s", curMAC, mac.String())
+		}
+		log.G(ctx).WithFields(log.Fields{"n": nSent + 1}).Debug("Sending unsolicited ARP/NA")
+		var errs []error
+		if arpSender != nil {
+			if err := arpSender.Send(); err != nil {
+				log.G(ctx).WithError(err).Warn("Failed to send unsolicited ARP")
+				errs = append(errs, err)
+			}
+		}
+		if naSender != nil {
+			if err := naSender.Send(); err != nil {
+				log.G(ctx).WithError(err).Warn("Failed to send unsolicited NA")
+				errs = append(errs, err)
+			}
+		}
+		return stderrors.Join(errs...)
+	}
+
+	// Send an initial message. If it fails, skip the resends.
+	if err := send(ctx, 0); err != nil {
+		return err
+	}
+	if i.advertiseAddrNMsgs == 1 {
+		return nil
+	}
+	// Don't clean up on return from this function, there are more ARPs/NAs to send.
+	stillSending = true
+
+	// Send the rest in the background.
+	go func() {
+		defer cleanup()
+		ctx, span := otel.Tracer("").Start(context.WithoutCancel(ctx), "libnetwork.osl.advertiseAddrs")
+		defer span.End()
+		ticker := time.NewTicker(i.advertiseAddrInterval)
+		defer ticker.Stop()
+		for nSent, nToSend := 1, i.advertiseAddrNMsgs; nSent < nToSend; nSent += 1 {
```

Comment:
Why are you copying `i.advertiseAddrNMsgs` into a local variable? If it's some ploy to prevent data races, I don't think it's effective.

Range-over-int can be used here; it's supported as of Go 1.22.
```suggestion
		for n := range i.advertiseAddrNMsgs - 1 {
```

Reply:
No ploy, it had that shape in the prev commit, before the config existed. But, sure.